### PR TITLE
Doc/qa process testing and dev

### PR DESCRIPTION
**Pull request by kats8**
## What

This PR adds several QA (quality assurance) files. Files/folders include:

A testing and dev document that covers QA related to the testing and development process.

A template folder containing Markdown testing sample templates (Test Strategy, Test Plan, Bug Report)

A quality assurance overview document which provides an overview/description and links to the other QA process files in the quality-assurance folder.

## Why

To provide testing & development process information and templates, and to organise them in a way that is useful and understandable to team members. QA processes are broken up into more manageable files based on context and the quality-assurance-overview is intended to draw a user's eye as a good place to begin if they are looking for an overview relating to QA info.

### CODE REVIEW

File: docs/processes/quality-assurance/git-contribution-guide.md [hunk 1]
```diff
@@ -212,7 +212,7 @@ git config --global core.editor "atom --wait"
 
 Ideally, we should avoid having developers working on the same code. It creates merge conflicts and hinders efficiency. If possible, try to break it down into small tasks so developers can work independently. In the worst-case scenario, if there is more than 1 person working on a feature branch:
 
-1. Please make sure you pull changes in the remove branch before starting your work.
+1. Please make sure you pull changes in the remote branch before starting your work.
```
Comment (tancnle) [inline docs/processes/quality-assurance/git-contribution-guide.md:215]: **praise:** ❤️

File: docs/processes/quality-assurance/quality-assurance-overview.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Guide to QA Processes
```
Comment (tancnle) [inline docs/processes/quality-assurance/quality-assurance-overview.md]: **suggestion:** If we use an acronym, I think we should spell it out on the first use on a page. We do not need to spell it out more than once on a page. When possible, we should try to avoid acronyms in headings. 

```suggestion
# Guide to Quality Assurance Processes
```

File: docs/processes/quality-assurance/quality-assurance-overview.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+[Why/What is QA?](#whywhat-is-qa)
+
+[Testing and Development](#testing-and-development)
+
+[Git Contribution Guidelines](#git-contribution-guidelines)
+
+## Why/What is QA?
+
+QA (Quality Assurance) is defined as a procedure to ensure the quality of software products or services provided to the customers by an organisation. Quality assurance focuses on improving the software development process and making it efficient and effective as per the quality standards defined for software products.
```
Comment (tancnle) [inline docs/processes/quality-assurance/quality-assurance-overview.md]: **suggestion:** We could introduce acronym at this point.

```suggestion
Quality Assurance (QA) is defined as a procedure to ensure the quality of software products or services provided to the customers by an organisation. QA focuses on improving the software development process and making it efficient and effective as per the quality standards defined for software products.
```

File: docs/processes/quality-assurance/quality-assurance-overview.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Guide to QA Processes
+
+[Why/What is QA?](#whywhat-is-qa)
+
+[Testing and Development](#testing-and-development)
+
+[Git Contribution Guidelines](#git-contribution-guidelines)
+
+## Why/What is QA?
```
Comment (tancnle) [inline docs/processes/quality-assurance/quality-assurance-overview.md]: **suggestion:**

```suggestion
## Why/What is Quality Assurance?
```

File: docs/processes/quality-assurance/quality-assurance-overview.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+- [Test-Driven Development](testing-and-dev.md#test-driven-development)
+
+- [Epics and User Stories](testing-and-dev.md#epics-and-user-stories)
+
+- [Testing Templates](testing-and-dev.md#testing-templates)
+
+## Git Contribution Guidelines
+
+The [Git Contributions Guidelines](git-contribution-guide.md) concerns processes relating to the git workflow, branching and commit formatting guidelines and draft pull requests.
```
Comment (tancnle) [inline docs/processes/quality-assurance/quality-assurance-overview.md]:
```suggestion
The [Git Contribution Guide](git-contribution-guide.md) concerns processes relating to the git workflow, branching and commit formatting guidelines and draft pull requests.
```

File: docs/processes/quality-assurance/templates/bug-report-template.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+## **Bug Report Template**
```
Comment (tancnle) [inline docs/processes/quality-assurance/templates/bug-report-template.md]: **nitpick:** Since this is a header, I think we don't need emphasis (ie. `**`). There is no visual difference 🤔

File: docs/processes/quality-assurance/templates/test-plan-template.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+## **Test Plan Template**
```
Comment (tancnle) [inline docs/processes/quality-assurance/templates/test-plan-template.md]: **nitpick:** Similar to suggestion above. Since this is a header, I think we don't need emphasis (ie. `**`). There is no visual difference 🤔

File: docs/processes/quality-assurance/templates/test-strategy-template.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+## **Test Strategy (Sample)**
```
Comment (tancnle) [inline docs/processes/quality-assurance/templates/test-strategy-template.md]: **nitpick:** Similar to suggestion above. Since this is a header, I think we don't need emphasis (ie. `**`). There is no visual difference 🤔

File: docs/processes/quality-assurance/templates/test-strategy-template.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+**Introduction**
+
+A high-level summary of the project
+
+**Example**
```
Comment (tancnle) [inline docs/processes/quality-assurance/templates/test-strategy-template.md]: **suggestion:** Small typo. `Examples`?

File: docs/processes/quality-assurance/templates/test-strategy-template.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
+## **Test Strategy (Sample)**
+
+**Introduction**
```
Comment (tancnle) [inline docs/processes/quality-assurance/templates/test-strategy-template.md]: **suggestion (non-blocking):** Could we use heading semantics for all below sections. 👇🏼 

```md
# Test Strategy

## Introduction

### Examples

## References

### Examples

...
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
+# QA Process - Testing and Development
+
+[Testing and Development](#testing-and-development
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]: **issue:** Missing closing bracket?

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+Thoth-Tech advocates for a [Test Driven Development](#test-driven-development) approach.
+
+But you might be wondering, before jumping into coding and planning tests for a new feature, _where do you begin_?
+
+## Feature development – where do I start?
+
+- Create [epics](#epics) for each feature for planned development
+- Identify [user stories](#user-stories) of the feature
+- Consider how best to mitigate risk or look for alternate solutions
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md:30]: **suggestion:** I think it is really important to articulate the problem, or why the changes need to be made. I suggest we add an extra clause to stress the importance of that step.

```
- Clearly articulate the problem and why the changes need to be made
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+You will also need to consider _how_ you are going to test your scenarios as part of this process – considering matters such as what tools will you use, what platforms tests will be performed on, any release processes, and what deliverables will you need to produce. This should be considered as part of your Test Strategy – a deliverable which should be produced as part of the QA process. Other documents that may be produced as part of the QA process include a Test Plan and Bug Reports.
+
+Templates have been provided for sample [Test Strategy](templates/test-strategy-template.md), [Test Plan](templates/test-plan-template.md) and [Bug Report](templates/bug-report-template.md) documents.
+
+While manual testing may be conducted initially, moving towards automated tests allow for better efficiency, ease of code checks and is the preferred method to allow automatic test incorporation into the deployment pipeline. Testing and processes, like many other in the Agile model, are not static, and can (and should) be improved on successive iterations. This may include iterating existing processes for testing release as well as automating more functionality.
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]:
```suggestion
While manual testing may be conducted initially, moving towards automated tests allows for better efficiency, ease of code checks and is the preferred method to allow automatic test incorporation into the deployment pipeline. Testing and processes, like many other in the Agile model, are not static, and can (and should) be improved on successive iterations. This may include iterating existing processes for testing releases as well as automating more functionality.
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+- [Refactoring Intro](#tbc-refactoring-intro)
+- [Waiting epic Template link](#tbc-template-link)
+- [Testing tools expansion](#tbc-testing-tools)
+
+# Testing and Development
+
+Testing should be built into the software development process.
+
+Thoth-Tech advocates for a [Test Driven Development](#test-driven-development) approach.
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]: **suggestion:** We could introduce the TDD acronym in the first use here. What do you think?

```suggestion
Thoth-Tech advocates for a [Test Driven Development](#test-driven-development) (TDD) approach.
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+
+Templates have been provided for sample [Test Strategy](templates/test-strategy-template.md), [Test Plan](templates/test-plan-template.md) and [Bug Report](templates/bug-report-template.md) documents.
+
+While manual testing may be conducted initially, moving towards automated tests allow for better efficiency, ease of code checks and is the preferred method to allow automatic test incorporation into the deployment pipeline. Testing and processes, like many other in the Agile model, are not static, and can (and should) be improved on successive iterations. This may include iterating existing processes for testing release as well as automating more functionality.
+
+## **How do I prove my code does what it's supposed to do?**
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]: **suggestion:** Could we drop the emphasis styles on headers 👇🏼 ? I don't think they add any visual difference.

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+Extract function refactoring is one common and intuitive approach to refactoring code: this involves comparing the code intention with the implementation. If the code is doing more than the intention, particularly if it is doing things that may be repeated elsewhere, it probably contains functions that should be split off.
+
+[Refactoring.com](http://refactoring.com/) provides useful reading that covers a number of refactoring techniques.
+
+**Further information about Test Driven Development**
+
+Deakin students have free access to LinkedIn Learning courses.
+
+A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things?u=2104084)
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]: **suggestion:** I think we can drop the `u` parameter as it is uniquely different for every user.

```suggestion
A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things).
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+You don't need to assess your code for refactoring on every test, but if your find your code is:
+
+- repeating itself (or repeating functionality),
+- confusing to follow, or
+- you are getting stuck,
+
+these are good signs that it might be time to refactor.
+
+**How do I refactor?**
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md:124]: **praise:** Great addition to the development guide ❤️

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+[Refactoring.com](http://refactoring.com/) provides useful reading that covers a number of refactoring techniques.
+
+**Further information about Test Driven Development**
+
+Deakin students have free access to LinkedIn Learning courses.
+
+A useful course on these general concepts is [Programming Foundations Test Driven Development](https://www.linkedin.com/learning/programming-foundations-test-driven-development-3/small-steps-to-great-things?u=2104084)
+
+Developers working with C++ may also find [Test Driven Development in C++](https://www.linkedin.com/learning/test-driven-development-in-c-plus-plus/welcome?autoplay=true&u=2104084) useful
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]:
```suggestion
Developers working with C++ may also find [Test Driven Development in C++](https://www.linkedin.com/learning/test-driven-development-in-c-plus-plus) useful.
```

File: docs/processes/quality-assurance/testing-and-dev.md [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+Further background on user stories can be found at https://www.atlassian.com/agile/project-management/user-stories
+
+## Testing Templates
+
+[Sample Test Strategy](templates/test-strategy-template.md))
+
+[Test Plan Template](templates/test-plan-template.md)
+
+[Bug Report Template](templates/bug-report-template.md)
```
Comment (tancnle) [inline docs/processes/quality-assurance/testing-and-dev.md]: **nitpick:** We can convert this to a bullet lists

```suggestion
- [Sample Test Strategy](templates/test-strategy-template.md))
- [Test Plan Template](templates/test-plan-template.md)
- [Bug Report Template](templates/bug-report-template.md)

```

**Review by tancnle — COMMENTED** [review]
Great effort so far @kats8 👍🏼 I love the details in the PR description. Thank you for adding it. I have left some comments/suggestions. Please let me know what you think? Apologies if I have suggested things that you are already working on already 🙏🏼.

File: docs/processes/quality-assurance/templates/test-strategy-template.md (hunk 9 above)
Comment (kats8) [reply docs/processes/quality-assurance/templates/test-strategy-template.md]: Hey Tan. Thanks for all the feedback. For this particular one, this subheading refers to this being  *an* example of *an* introduction, so I believe Example is correct here :)

File: docs/processes/quality-assurance/templates/test-strategy-template.md (hunk 10 above)
Comment (kats8) [reply docs/processes/quality-assurance/templates/test-strategy-template.md]: I like this suggestion. Thanks @tancnle  

Incorporating 👍

File: docs/processes/quality-assurance/testing-and-dev.md (hunk 15 above)
Comment (kats8) [reply docs/processes/quality-assurance/testing-and-dev.md]: Ah... So it seems this _is_ making a visual difference on (at least some) headers - on my VS preview, anyway. The headings are only a little larger, but not bold without it - second image is removing the emphasis. Smaller headings especially seem to get lost a bit without it to my eye? @tancnle 
image
image

File: docs/processes/quality-assurance/testing-and-dev.md (hunk 16 above)
Comment (kats8) [reply docs/processes/quality-assurance/testing-and-dev.md]: Thanks for picking that up, @tancnle !

File: docs/processes/quality-assurance/testing-and-dev.md (hunk 19 above)
Comment (kats8) [reply docs/processes/quality-assurance/testing-and-dev.md]: Done - with a new intro sentence, as indented bullets looked a little off as the only thing under a heading.

**Review by ginge270 — APPROVED** [review]
